include/range/v3/view/cycle.hpp

```diff
@@ -0,0 +1,259 @@
+/// \file cycle.hpp
+// Range v3 library
+//
+//  Copyright Eric Niebler 2013-2015
+//  Copyright Gonzalo Brito Gadeschi 2015
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_CYCLE_HPP
+#define RANGES_V3_VIEW_CYCLE_HPP
+#include <iostream>
+#include <utility>
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/size.hpp>
+#include <range/v3/distance.hpp>
+#include <range/v3/begin_end.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/view_facade.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/utility/iterator.hpp>
+#include <range/v3/utility/static_const.hpp>
+
+namespace ranges
+{
+
+    inline namespace v3
+    {
+
+        /// \addtogroup group-views
+        ///@{
+        template <typename Rng>
+        struct cycled_view : view_facade<cycled_view<Rng>, infinite>
+        {
+
+        private:
+            CONCEPT_ASSERT(ForwardRange<Rng>());
+            friend range_access;
+            Rng rng_;
+
+        public:
+            cycled_view() = default;
+            explicit cycled_view(Rng rng) : rng_(std::move(rng))
+            {
+            }
+
+        private:
+            template <bool IsConst> struct cursor;
+            template <bool IsConst> struct sentinel;
+
+            template <bool IsConst>
+            struct cursor
+            {
+            private:
+                template <typename T>
+                using constify_if = meta::apply<meta::add_const_if_c<IsConst>, T>;
+                using cycled_view_t = constify_if<cycled_view>;
+            public:
+                // The difference type of the cycled view does not depend
+                // on the difference type of the underlying range.
+                // TODO: more meaningful value?
+                // E.g. max<ptrdiff_t, range_difference_t<Rng>>?
+                using difference_type = std::ptrdiff_t;
+                using reference = range_reference_t<constify_if<Rng>>;
+                using single_pass = std::false_type;
+
+            private:
+                friend struct sentinel<IsConst>;
+                cycled_view_t *rng_;
+                // TODO: RandomAccessRange's don't need to store the iterator
+                // since it can be obtained from `begin` in O(1).
+                // Note: Bidirectional and Forward ranges need to store the
+                // iterator for `current` to be O(1).
+                range_iterator_t<constify_if<Rng>> it_;
+                difference_type pos_;
```

Comment:
Agreed, that should be achieved via the `done` method on the cursor always returning `false` right?

In the example above `end` is the end iterator of a `take` view (over a cyclic range). 

Basically the "problem" `pos_` solved was that if my sequence seq has N elements, I cycle it `auto rng = seq | view::cycled`, and then I take the slice `[begin(rng), begin(rng) + 3*N)`, the distance of that slice should be `3*N`. If pointing to the same element is sufficient for two iterators to compare equal, the slice `[begin(rng), begin(rng) + 3*N)` might have distance zero (since `begin(rng)  + M * N` points to `begin(rng)`).